Allow low-level use of Server

JoseKilo opened the pull request:
Close: #265 

Basic idea is being able to do something like this:

```
config = Config(app=make_app())
server = Server(config)
loop = asyncio.get_event_loop()
loop.run_until_complete(server.run())
```

This will allow to run other tasks in the same asyncio loop.

JoseKilo committed:
Allow low-level use of Server

JoseKilo commented on uvicorn/loops/uvloop.py:
```diff
@@ -4,6 +4,5 @@
 
 
 def uvloop_setup():
-    asyncio.get_event_loop().close()
```
Not sure if this is actually needed. If I keep the line, it complains with `RuntimeError: Cannot close a running event loop` (using uvloop only).

JoseKilo commented on uvicorn/main.py:
```diff
@@ -256,26 +256,37 @@ def __init__(self, config):
 
-        self.install_signal_handlers()
+        if config.install_signal_handlers:
+            self.install_signal_handlers()
```
This check was missing, but the config property was there

lovelydinosaur commented:
Okay, so I think we might want to get from A to B step by step here, as I think there's some related changes to make along the way.

First step: Let's *just* deal with splitting out the `Config.setup_event_loop`

I think we probably want to drop `loop_instance` on the config, and instead use `asyncio.get_event_loop()` everywhere instead. The only place where we can't do that is on the protocol classes, because for the sake of testing we want to be able to inject mock loop implementations there.

So, I think step one here is probably:

* Split out `Config.setup_event_loop()`
* Drop the `loop_instance` on the config completely.
* Use `asyncio.get_event_loop()` everywhere.
* In the protocol classes add an argument `_loop = None` on init, and use `self.loop = _loop or asyncio.get_event_loop()` - the test cases can then explicitly set the loop on the protocol.

Once we've got that in, then I think we should tackle the rest of the `main()` vs. `async run()` refactoring.

Does that seem reasonable?

JoseKilo commented:
Yes, it makes sense. Let's start with the first one: #301

JoseKilo committed:
Remove `install_signal_handlers` option from config

JoseKilo commented on uvicorn/config.py:
```diff
@@ -83,7 +83,6 @@ def __init__(
         timeout_keep_alive=5,
         timeout_notify=30,
         callback_notify=None,
-        install_signal_handlers=True,
```
Not sure if we want to remove it from the arguments (wouldn't be fully backwards compatible)

lovelydinosaur replied:
It's not documented API, so we're good.
(We were just using it internally)

JoseKilo commented:
I have rebased on top of master and left the `install_signal_handlers` change out.

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -401,12 +407,15 @@ def install_signal_handlers(self):
             self.should_exit = True
 
+    def handle_exit(self, sig, frame):
+        self.stop()
```
Let's not add this extra indirection.

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -256,28 +256,34 @@ def __init__(self, config):
 
+    @property
+    def loop(self):
+        return asyncio.get_event_loop()
```
I think we should probably drop this, and just use `loop = asyncio.get_event_loop()` inside any methods that need it.

lovelydinosaur commented on uvicorn/main.py:
```diff
@@ -256,28 +256,34 @@ def __init__(self, config):
-    def run(self):
+    def main(self):
+        self.config.setup_event_loop()
+
```
This newline probably isn't meant to be here.

lovelydinosaur reviewed (changes requested):
Looking really good!

I've put some minor comments in place to address. Other thing is: are we happy we have the naming right here?

Right now we have:

* `def uvicorn.run()`
* `def Server.main()`
* `async def Server.run()`

Which is a bit asymmetrical. Perhaps we should swap `run` and `main` here??? Or something else?

lovelydinosaur commented:
Or perhaps:

`uvicorn.run()`
`server.run()`
`await server.serve()`

lovelydinosaur commented:
Not sure exactly, but I guess it probably makes sense for the blocking server method name to match the blocking function name.

JoseKilo committed:
Avoid extra indirection

JoseKilo committed:
Avoid extra property

JoseKilo committed:
Remove newline

JoseKilo committed:
Better naming scheme

uvicorn.run -> Server.run -> Server.serve

Blocking function matches blocking server method.
Blocking server method awaits async server method.

JoseKilo committed:
Style fix

JoseKilo commented:
Going with `uvicorn.run -> Server.run -> Server.serve`.

That way the previous Server API doesn't change, and the tests don't need to change.

lovelydinosaur commented:
Lovely, lovely stuff here! 🎉

I'm going to tweak the `__init__` imports to make `Server` and `Config` available from the main module. And also roll a new release.

At which point we will have the following:

Standard entry point:

```python
import uvicorn

uvicorn.run(app, **kwargs)
```

Run, blocking the current thread.

```python
import uvicorn

config = uvicorn.Config(app, **kwargs)
server = uvicorn.Server(config)
server.run()
```

Run using the event loop explictly.

```python
import asyncio
import uvicorn

config = uvicorn.Config(app, **kwargs)
server = uvicorn.Server(config)
loop = asyncio.get_event_loop()
loop.run_until_complete(server.serve())
```

In due course we should start documenting up the programatic interface.

JoseKilo commented:
That's great ! Thank you for your help :)
